doc/stdenv/stdenv.chapter.md

```diff
@@ -305,6 +305,61 @@ The script will be usually run from the root of the Nixpkgs repository but you s
 
 For information about how to run the updates, execute `nix-shell maintainers/scripts/update.nix`.
 
+### Recursive attributes in `mkDerivation`
+
+If you pass a function to `mkDerivation`, it will receive as its argument the final arguments, considering use of `overrideAttrs`. For example:
+
+```nix
+mkDerivation (finalAttrs: {
+  pname = "hello";
+  withFeature = true;
+  configureFlags =
+    lib.optionals finalAttrs.withFeature ["--with-feature"];
+})
+```
+
+Note that this does not use the `rec` keyword to reuse `withFeature` in `configureFlags`.
+The `rec` keyword works at the syntax level and is unaware of overriding.
+
+Instead, the definition references `finalAttrs`, allowing users to change `withFeature`
+consistently with `overrideAttrs`.
+
+`finalAttrs` also contains the attribute `public`, which represents the final package,
+including the output paths, etc.
```

Comment:
What happens if the derivation attributes included a field called `public`?

What happens if I write something like `drv.overrideAttrs (final: final)`? Won't that turn the package into an input of itself?

Reply:
The `mkDerivation` caller or overrider can still set `public`, but the overrider can only read that value back from `previousAttrs`; not `finalAttrs`. `previousAttrs` is also the legacy one, so this won't break derivations that already have a `public` attr, which would be out-of-tree ones.

> `drv.overrideAttrs (final: final)`

This parameter shouldn't be called `final`, as a one-argument `overrideAttrs` behaves as it always did: `previousAttrs` (or `old`). It is equal to `drv`.

> something like `drv.overrideAttrs (final: final)`

Just like with overlays, `drv.overrideAttrs (finalAttrs: previousAttrs: finalAttrs)` produces a detectable infinite recursion.

I've managed to insert a custom error message, even if it's a bit hacky. Here's a somewhat realistic attempt of trying to define an attr only when another attr is defined. This is fundamentally impossible in Nix (as it currently exists; I believe @Infinisil tried to change this once. I don't know if that project has succeeded.), because the attributes part of the attrset is a strict data structure. 

```
nix-repl> (stdenv.mkDerivation { name = "hi"; public = "hi"; }).overrideAttrs (finalAttrs: previousAttrs: lib.optionalAttrs (finalAttrs?foo) { bar = "Only if we'll have foo for sure!"; }) 
error: infinite recursion encountered

       at /home/user/h/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:27:22:

           26|       # An infinite recursion here can be caused by having the attribute names of expression `e` in .overrideAttrs(finalAttrs: previousAttrs: e) depend on `finalAttrs`. Only the attribute values of `e` can depend on `finalAttrs`.
           27|       args = rattrs (args // { inherit public; });
             |                      ^
           28|       #              ^^^^
```

Reply:
Infinite recursion involving `builtins.derivation`. I feel like this is more unlikely and it prints a somewhat informative error message:

```
$ nix-instantiate --expr 'with import ./. {}; (stdenv.mkDerivation { name = "hi"; public = "hi"; }).overrideAttrs (finalAttrs: previousAttrs: { bar = finalAttrs.public; })' --show-trace
error: infinite recursion encountered

       at //builtin/derivation.nix:19:19:
       … while evaluating the attribute 'bar' of the derivation 'hi'

       at /home/user/h/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:270:7:

          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let

```

Reply:
A new way of self-bootstrapping. Wouldn't recommend because this doesn't have cross compilation splicing support, but it's interesting. Maybe cross compilation splicing could be done via a function similar to `overrideAttrs`, invoked by the cross compilation code. I've played around with the idea of moving the code that improves the output attrs into a postcomposed function that would defined in the fixpoint to be overridable. Perhaps that would allow it, but the interaction between the `overrideAttrs` and such a function isn't easy to get right.

Anyway, for the curious:

```
stdenv.mkDerivation(finalAttrs: {
  name = "bootstrapper";
  nativeBuildInputs = [ finalAttrs.bootPkg ];
  bootPkg = finalAttrs.public.overrideAttrs (_: { bootPkg = null; });
  dontUnpack = true;
  installPhase = ''
    mkdir -p $out/bin
    if test -n "''${bootPkg}"; then
      $bootPkg/bin/boot "world" >$out/msg
    fi
    dollar='$'
    cat >$out/bin/boot <<EOF
    #!${runtimeShell}
    echo "Hello ''${dollar}1!"
    if test -e $out/msg; then
      echo "also"
      cat $out/msg
    fi
    EOF
    chmod a+x $out/bin/boot
  '';
})
```

```
$ ./result/bin/boot there
Hello there!
also
Hello world!
```

Reply:
Sorry yes I did mean `drv.overrideAttrs (old: old)`. I wasn't reading carefully enough though to realize where the `public` attr went.

In any case, what happens if a derivation already has a `public` attr? Does `finalAttrs.public` just hide that original `public` attr? Can you say `finalAttrs.public.public` to get at it?

Reply:
> In any case, what happens if a derivation already has a `public` attr? Does `finalAttrs.public` just hide that original `public` attr?

Yes

> Can you say `finalAttrs.public.public` to get at it?

Yes, `mkDerivation` returns most if not all of its arguments, so you can.

```
nix-repl> (stdenv.mkDerivation (finalAttrs: { public = true; })).public                          
true

nix-repl> (stdenv.mkDerivation (finalAttrs: { public = true; viaFinal = finalAttrs.public.public; })).viaFinal
true

nix-repl> p = stdenv.mkDerivation (finalAttrs: { public = true; })

nix-repl> (p.overrideAttrs (previousAttrs: { via = previousAttrs.public; })).via
true

nix-repl> (p.overrideAttrs (finalAttrs: previousAttrs: { via = finalAttrs.public.public; })).via
true
```

Reply:
Ok thank you, that satisfies my questions.